release/releaser.go

```diff
@@ -45,539 +36,318 @@ func NewReleaser(
-	stage = metrics.NewTimer(base.With("stage", "fetch_platform_services"))
-
-	services, err := getServices.SelectServices(inst)
+func sendNotifications(inst *instance.Instance, executeErr error, results flux.ReleaseResult, job *jobs.Job) error {
+	cfg, err := inst.GetConfig()
 	if err != nil {
-		return nil, errors.Wrap(err, "fetching platform services")
-	}
-	if len(services) == 0 {
-		res = append(res, r.releaseActionPrintf("No selected services found. Nothing to do."))
-		return res, nil
+		if executeErr == nil {
+			return errors.Wrap(err, "sending notifications")
+		}
+		return executeErr
 	}
 
-	stage.ObserveDuration()
-	stage = metrics.NewTimer(base.With("stage", "finalize"))
-
-	res = append(res, r.releaseActionPrintf(msg))
-	res = append(res, r.releaseActionClone())
-
-	ids := []flux.ServiceID{}
-	for _, service := range services {
-		res = append(res, r.releaseActionFindPodController(service.ID))
-		ids = append(ids, service.ID)
+	status := flux.ReleaseStatusSuccess
+	if executeErr != nil {
+		status = flux.ReleaseStatusFailed
+	}
+	// Filling this from the job is a temporary migration hack. Ideally all
+	// the release info should be stored on the release object in a releases
+	// table, and the job should really just have a pointer to that.
+	err = notifications.Release(cfg, flux.Release{
+		ID:        flux.ReleaseID(job.ID),
+		CreatedAt: job.Submitted,
+		StartedAt: job.Claimed,
+		// TODO: fetch the job and look this up so it matches
+		// (which must be done after completing the job)
+		EndedAt:  time.Now().UTC(),
+		Done:     true,
+		Priority: job.Priority,
+		Status:   status,
+		Log:      job.Log,
+
+		Spec:   flux.ReleaseSpec(job.Params.(jobs.ReleaseJobParams)),
+		Result: results,
+	}, executeErr)
+	if err != nil {
+		if executeErr == nil {
+			return errors.Wrap(err, "sending notifications")
+		}
 	}
-	res = append(res, r.releaseActionReleaseServices(ids, msg))
-	return res, nil
+	return nil
 }
 
-func (r *Releaser) execute(inst *instance.Instance, actions []ReleaseAction, kind flux.ReleaseKind, job *jobs.Job, updateJob func(string, ...interface{})) (err error) {
-	rc := NewReleaseContext(inst)
-	defer rc.Clean()
-
-	defer func() {
-		cfg, err2 := inst.GetConfig()
-		if err2 != nil {
-			if err == nil {
-				err = errors.Wrap(err2, "sending notifications")
+// Take the spec given in the job, and figure out which services are
+// in question based on the running services and those defined in the
+// repo. Fill in the release results along the way.
+func selectServices(rc *ReleaseContext, spec *flux.ReleaseSpec, results flux.ReleaseResult, logStatus statusFn) ([]*ServiceUpdate, error) {
+	conf, err := rc.Instance.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	lockedSet := LockedServices(conf)
+
+	excludedSet := flux.ServiceIDSet{}
+	excludedSet.Add(spec.Excludes)
+
+	// For backwards-compatibility, there's two fields: ServiceSpec
+	// and ServiceSpecs. An entry in ServiceSpec takes precedence.
```

Comment:
Could quarantine it to the http/transport, with a
```
type LegacyReleaseSpec struct {
  flux.ReleaseSpec
  ServiceSpec flux.ServiceSpec `json:"serviceSpec"`
}
```
then convert it or w/e. maybe that is worth doing... :thinking: